File: patterns/rag/code/python/rag.py
"""
RAG (Retrieval-Augmented Generation) — Retrieve relevant context, then generate.

Two phases:
  Ingestion (offline): chunk documents → embed → store in vector store
  Query (online):      embed query → retrieve top-k chunks → generate answer

Design doc:  ../../design.md
Overview:    ../../overview.md
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol


# ── Interfaces ────────────────────────────────────────────────────────────────

class LLM(Protocol):
    def generate(self, messages: list[dict]) -> str: ...


class Embedder(Protocol):
    """Convert text to a float vector. Use any embedding model."""
    def embed(self, text: str) -> list[float]: ...


class VectorStore(Protocol):
    """Minimal vector store interface."""
    def add(self, id: str, embedding: list[float], text: str, metadata: dict) -> None: ...
    def search(self, embedding: list[float], top_k: int) -> list["Chunk"]: ...


# ── Core types ────────────────────────────────────────────────────────────────

@dataclass
class Chunk:
    id: str
    text: str
    score: float = 0.0
    metadata: dict = field(default_factory=dict)


@dataclass
class RAGResult:
    answer: str
    chunks_used: list[Chunk] = field(default_factory=list)
    query: str = ""


# ── In-memory vector store (for testing) ─────────────────────────────────────

class InMemoryVectorStore:
    """
    Simple dot-product similarity store.
    Replace with Chroma, Pinecone, Weaviate, etc. in production.
    """

    def __init__(self):
        self._items: list[tuple[str, list[float], str, dict]] = []

    def add(self, id: str, embedding: list[float], text: str, metadata: dict) -> None:
        self._items.append((id, embedding, text, metadata))

    def search(self, embedding: list[float], top_k: int) -> list[Chunk]:
        def dot(a: list[float], b: list[float]) -> float:
            return sum(x * y for x, y in zip(a, b))

        scored = [
            Chunk(id=id_, text=text, score=dot(embedding, emb), metadata=meta)
            for id_, emb, text, meta in self._items
        ]
        scored.sort(key=lambda c: c.score, reverse=True)
        return scored[:top_k]


# ── RAG pipeline ──────────────────────────────────────────────────────────────

ANSWER_PROMPT = """\
Answer the question using only the provided context. If the context does not
contain enough information, say so clearly.

Context:
{context}

Question: {question}"""


class RAGPipeline:
    """
    Retrieval-Augmented Generation pipeline.

    Ingestion: call pipeline.ingest(documents) once (or incrementally).
    Query:     call pipeline.query(question) at request time.
    """

    def __init__(
        self,
        llm: LLM,
        embedder: Embedder,
        vector_store: VectorStore | None = None,
        top_k: int = 3,
        chunk_size: int = 500,
        chunk_overlap: int = 50,
        system: str = "",
    ):
        self.llm = llm
        self.embedder = embedder
        self.store = vector_store or InMemoryVectorStore()
        self.top_k = top_k
        self.chunk_size = chunk_size
        self.chunk_overlap = chunk_overlap
        self.system = system
        self._chunk_counter = 0

    def _split(self, text: str) -> list[str]:
        """Split text into overlapping chunks."""
        chunks: list[str] = []
        start = 0
        while start < len(text):
            end = start + self.chunk_size
            chunks.append(text[start:end].strip())
            start += self.chunk_size - self.chunk_overlap
        return [c for c in chunks if c]

    def ingest(self, documents: list[str], metadata: list[dict] | None = None) -> int:
        """
        Chunk, embed, and store documents.
        Returns the number of chunks added.
        """
        metadata = metadata or [{}] * len(documents)
        added = 0
        for doc, meta in zip(documents, metadata):
            for chunk_text in self._split(doc):
                chunk_id = f"chunk_{self._chunk_counter}"
                self._chunk_counter += 1
                embedding = self.embedder.embed(chunk_text)
                self.store.add(chunk_id, embedding, chunk_text, meta)
                added += 1
        return added

    def query(self, question: str) -> RAGResult:
        # Embed the question and retrieve relevant chunks
        q_embedding = self.embedder.embed(question)
        chunks = self.store.search(q_embedding, top_k=self.top_k)

        # Build context from retrieved chunks
        context = "\n\n---\n\n".join(
            f"[Source {i + 1}]\n{chunk.text}" for i, chunk in enumerate(chunks)
        )

        # Generate grounded answer
        messages: list[dict] = []
        if self.system:
            messages.append({"role": "system", "content": self.system})
        messages.append({
            "role": "user",
            "content": ANSWER_PROMPT.format(context=context, question=question),
        })

        answer = self.llm.generate(messages)
        return RAGResult(answer=answer, chunks_used=chunks, query=question)


# ── Example ───────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    import hashlib

    class MockEmbedder:
        """Deterministic fake embeddings based on text hash."""
        def embed(self, text: str) -> list[float]:
            h = int(hashlib.md5(text.encode()).hexdigest(), 16)
            return [(h >> i & 0xFF) / 255.0 for i in range(8)]

    class MockLLM:
        def generate(self, messages: list[dict]) -> str:
            return f"Based on the context: {messages[-1]['content'][:80]}..."

    pipeline = RAGPipeline(
        llm=MockLLM(),
        embedder=MockEmbedder(),
        top_k=2,
        chunk_size=200,
    )

    # Ingest documents
    n = pipeline.ingest([
        "ReAct is a prompting technique that combines chain-of-thought reasoning with action execution.",
        "RAG stands for Retrieval-Augmented Generation. It retrieves relevant documents before generating.",
        "Plan and Execute separates planning from execution. The planner creates a full plan upfront.",
    ])
    print(f"Ingested {n} chunks")

    # Query
    result = pipeline.query("What is RAG and how does it work?")
    print(f"Retrieved {len(result.chunks_used)} chunks")
    print(f"Answer: {result.answer[:120]}")

File: patterns/rag/code/typescript/vercel-ai-sdk/rag.ts
/**
 * RAG — Vercel AI SDK variant.
 *
 * Pattern: RAG (Retrieval-Augmented Generation: chunk + embed + retrieve, then generate).
 * Framework: Vercel AI SDK (ai ^4.0.0) + @ai-sdk/anthropic (^1.0.0).
 * Idioms: retrieve happens in app code (not as a model-side tool), and the
 *   retrieved chunks are inlined into the system prompt before generateText
 *   produces the grounded answer. Matches the dominant Vercel AI SDK RAG shape.
 * Design doc: ../../../design.md (the framework-agnostic ../../python/rag.py
 *   walks the chunk → embed → store → retrieve → generate cycle with the
 *   same three seed documents and the same query).
 *
 * Install:  pnpm add ai @ai-sdk/anthropic
 *           (or via tsx-on-demand: npx tsx rag.ts)
 * Run:      ANTHROPIC_API_KEY=... npx tsx rag.ts
 *
 * Why retrieve outside the model call: the Vercel AI SDK's strength is the
 * thin generateText/streamText surface. Wrapping retrieval as a tool the model
 * has to call costs an extra round trip with no quality gain for the simple
 * case. Reach for the agent-with-retriever-tool shape only when the model
 * needs to decide *whether* to retrieve — see the recipe at
 * docs/recipes/docs-rag-qa.md for the agentic variant.
 *
 * Note: ESM only (matches the canonical TypeScript project layout).
 * tsconfig snippet: { "module": "ESNext", "moduleResolution": "Bundler", "target": "ES2022" }
 */

import { createHash } from "node:crypto";

import { anthropic } from "@ai-sdk/anthropic";
import { generateText } from "ai";

// ── Types ────────────────────────────────────────────────────────────────────

interface Chunk {
  id: string;
  text: string;
  score: number;
  metadata: Record<string, unknown>;
}

interface RAGResult {
  answer: string;
  chunksUsed: Chunk[];
  query: string;
}

interface Embedder {
  embed(text: string): number[];
}

// ── Deterministic mock embedder (matches the Python sibling) ─────────────────

class HashEmbedder implements Embedder {
  embed(text: string): number[] {
    // 8-dim vector seeded from the md5 of the text. Same shape as the
    // MockEmbedder in ../../python/rag.py so the two siblings can be
    // diffed at the chunk-score level.
    const digest = createHash("md5").update(text).digest("hex");
    const big = BigInt(`0x${digest}`);
    const vec: number[] = [];
    for (let i = 0; i < 8; i++) {
      const shifted = (big >> BigInt(i * 8)) & 0xffn;
      vec.push(Number(shifted) / 255);
    }
    return vec;
  }
}

// ── In-memory vector store ───────────────────────────────────────────────────
// Replace with pgvector (rides the relational database already in the stack),
// Qdrant, or Chroma in production.

class InMemoryVectorStore {
  private items: Array<{ id: string; embedding: number[]; text: string; metadata: Record<string, unknown> }> = [];

  add(id: string, embedding: number[], text: string, metadata: Record<string, unknown>): void {
    this.items.push({ id, embedding, text, metadata });
  }

  search(embedding: number[], topK: number): Chunk[] {
    const dot = (a: number[], b: number[]): number => a.reduce((sum, x, i) => sum + x * b[i], 0);
    return this.items
      .map((item) => ({
        id: item.id,
        text: item.text,
        score: dot(embedding, item.embedding),
        metadata: item.metadata,
      }))
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);
  }
}

// ── Pipeline ────────────────────────────────────────────────────────────────

interface RAGOptions {
  topK?: number;
  chunkSize?: number;
  chunkOverlap?: number;
  system?: string;
}

const DEFAULT_SYSTEM =
  "You answer questions from the provided context. If the context does not " +
  "contain the answer, say so plainly instead of guessing.";

class RAGPipeline {
  private store = new InMemoryVectorStore();
  private chunkCounter = 0;

  constructor(
    private readonly embedder: Embedder,
    private readonly model = anthropic("claude-haiku-4-5"),
    private readonly opts: Required<RAGOptions> = {
      topK: 2,
      chunkSize: 200,
      chunkOverlap: 50,
      system: DEFAULT_SYSTEM,
    },
  ) {}

  private split(text: string): string[] {
    const chunks: string[] = [];
    let start = 0;
    while (start < text.length) {
      const end = start + this.opts.chunkSize;
      const slice = text.slice(start, end).trim();
      if (slice) chunks.push(slice);
      start += this.opts.chunkSize - this.opts.chunkOverlap;
    }
    return chunks;
  }

  ingest(documents: string[], metadata: Array<Record<string, unknown>> = []): number {
    let added = 0;
    documents.forEach((doc, i) => {
      const meta = metadata[i] ?? {};
      for (const text of this.split(doc)) {
        const id = `chunk_${this.chunkCounter++}`;
        this.store.add(id, this.embedder.embed(text), text, meta);
        added++;
      }
    });
    return added;
  }

  async query(question: string): Promise<RAGResult> {
    const qEmbedding = this.embedder.embed(question);
    const chunks = this.store.search(qEmbedding, this.opts.topK);

    const context = chunks
      .map((c, i) => `[Source ${i + 1}]\n${c.text}`)
      .join("\n\n---\n\n");

    const userPrompt =
      `Answer the question using only the provided context. If the context does ` +
      `not contain enough information, say so clearly.\n\n` +
      `Context:\n${context}\n\nQuestion: ${question}`;

    const result = await generateText({
      model: this.model,
      system: this.opts.system,
      prompt: userPrompt,
    });

    return { answer: result.text, chunksUsed: chunks, query: question };
  }
}

// ── Smoke runner ────────────────────────────────────────────────────────────

const SEED_DOCUMENTS = [
  "ReAct is a prompting technique that combines chain-of-thought reasoning with action execution.",
  "RAG stands for Retrieval-Augmented Generation. It retrieves relevant documents before generating.",
  "Plan and Execute separates planning from execution. The planner creates a full plan upfront.",
];

async function main(): Promise<void> {
  if (!process.env.ANTHROPIC_API_KEY) {
    console.log("Skipping smoke run — set ANTHROPIC_API_KEY to exercise the real generate path.");

    // Run ingestion + retrieval without the model so the dispatcher contract
    // is still exercisable offline.
    const pipeline = new RAGPipeline(new HashEmbedder());
    const n = pipeline.ingest(SEED_DOCUMENTS);
    console.log(`Ingested ${n} chunks (offline).`);
    return;
  }

  const pipeline = new RAGPipeline(new HashEmbedder());
  const n = pipeline.ingest(SEED_DOCUMENTS);
  console.log(`Ingested ${n} chunks`);

  const result = await pipeline.query("What is RAG and how does it work?");
  console.log(`Retrieved ${result.chunksUsed.length} chunks`);
  console.log(`Answer: ${result.answer.slice(0, 120)}`);
}

// ESM entry-point check.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

export { RAGPipeline, InMemoryVectorStore, HashEmbedder };
export type { Chunk, RAGResult, Embedder, RAGOptions };
